### PR TITLE
Lodash: Add missing use cases

### LINKED ISSUE #20741: [@types/lodash] differenceBy types are more strict than code suggests
If you know how to fix the issue, make a pull request instead.

- [X] I tried using the `@types/lodash` package and had problems.
- [ ] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [X] I'm using Typescript 2.3.4 (latest dependency available for my Angular project)
- [X] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [ ] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @bczengel @chrootsu @stepancar @ericanderson @aj-r @Ailrun @e-cloud

I have a couple of different classes that use the same semantic property name. Prior to the 4.14.78 release, I could do the following:

```
class A {
  constructor(myName) {
    this.name = myName;
  }
}

class B {
  constructor(myName, age) {
    this.name = myName;
    this.age = age;
  }
}

let a1 = new A('a1');
let a2 = new A('a2');
let b1 = new B('a1', 30);

let diff = _.differenceBy([a1, a2], [b1], 'name');
// result: [{name: a2}]
```

Now, the type definition prevents me from doing this because class A and B are not related. It doesn't appear to be a restriction in lodash code itself. Should the type definition be this strict?

### PR DESCRIPTION
**Pull request by aj-r**
Fixes #20741. I've added support for some subtle use cases that broke with #19356, specifically for `differenceBy` and `flatMap`.

For the context of these changes, see:

- https://github.com/DefinitelyTyped/DefinitelyTyped/issues/20741#issuecomment-338254190
- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19356#issuecomment-338194048

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://lodash.com/docs/4.17.4#differenceBy, https://lodash.com/docs/4.17.4#flatMap
- [x] Increase the version number in the header if appropriate. **(not appropriate)**
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. (already exists)

### CODE REVIEW

**Comment by karptonite** [comment]
Note that this problem also affects `differenceWith`, perhaps even more, since differenceWith takes a callback, and there is no reason to think that the callback couldn't take different types for the first and second arguments.

**Comment by aj-r** [comment]
@andy-ms Any idea how to fix the GC error in Travis? I get the same error when running `npm run lint lodash` locally but have no idea how to fix it. It always happens when testing with TS 2.3.

**Comment by andy-ms** [comment]
Looks like we've fixed some bugs since ts2.3. Also, tests will fail in ts2.3 because type inference has been improved. You could just specify `// TypeScript Version: 2.4` in the header.

**Comment by aj-r** [comment]
@bczengel @chrootsu @stepancar @ericanderson @Ailrun @e-cloud Checks are now passing. Could you review this PR?

**Comment by aj-r** [comment]
I guess @thorn0 could also review this, since he's an author now.

**Comment by thorn0** [comment]
Reverting the changes one by one and repeatedly running lint, you can find what breaks TS 2.3. That's what I did with my PR.

File: types/lodash/index.d.ts
Comment (thorn0) [inline types/lodash/index.d.ts]: Why the return type is `any`?

Comment (aj-r) [reply types/lodash/index.d.ts]: Lodash looks for any truthy or falsy value in the return value, so we shouldn't restrict it to just booleans. For example, see this comparator: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/20795/files#diff-6a82a43b7b79f07644334583c1add52aR671. `boolean | undefined` should be an allowed return type because lodash will treat `undefined` as a falsy value.

Comment (andy-ms) [reply types/lodash/index.d.ts]: I don't buy that logic -- if something will be treated as a boolean it should be typed as a `boolean`. `Math.abs("-2")` will work at runtime but isn't allowed by TypeScript.

Comment (aj-r) [reply types/lodash/index.d.ts]: Returning a non-boolean value in callbacks is a common use case that people have opened issues for in the past - see #18956. If we force a boolean return value in comparators, then for consistency I think we should also force boolean return values in functions like `filter`, `some`, and `every`. But that would break #18956.

Comment (thorn0) [reply types/lodash/index.d.ts]: Let's break it. TypeScript's philosophy in regard to this kind of forgiveness and implicit conversions seems to be clear and simple.

File: types/lodash/index.d.ts
Comment (aj-r) [reply types/lodash/index.d.ts]: Good catch, Thanks!

File: types/lodash/index.d.ts
Comment (aj-r) [reply types/lodash/index.d.ts]: What if we allow the return type to be `boolean | null | undefined`? That would support the use case in #18956, but also prevent the user from returning bad things like strings, objects, or functions.

Comment (aj-r) [reply types/lodash/index.d.ts]: Then again, maybe it's not a big deal to force users to put `!!` in front of anything that would otherwise be undefined.

Comment (thorn0) [reply types/lodash/index.d.ts]: `boolean | null | undefined` looks like a good trade-off. In particular, there will still be an error if the user forgets to `return` a value (`x => x` vs `x => { x }`) due to the fact that TS makes a distinction between `undefined` and `void`.

In Java and, even the condition in `if` statement must be of boolean type.

Comment (thorn0) [reply types/lodash/index.d.ts]: In Java and the like, even conditions in `if` statements must be of boolean type, which was considered a step forward to type safety. Everyone seems to have been okay with this design decision.

Comment (thorn0) [reply types/lodash/index.d.ts]: This made me think, what if change the type of `ValueIteratee` for functions like `differenceBy` from the current shape `(value: T) => any` to `(value: T) => {} | undefined | null`. If I'm not mistaken, `{} | undefined | null` means "`any` except `void`". And an iteratee callback that never returns a value is definitely a mistake.

**Comment by thorn0** [comment]
The rest of the PR looks good. Also you were going to fix `isFunction`, weren't you?

**Comment by aj-r** [comment]
Looks like someone else is fixing `isFunction`: #21015

**Comment by aj-r** [comment]
I've discovered that if you run:

`node node_modules\dtslint\bin\index.js types lodash`

instead of:

`npm run lint lodash`

then it will tell you the actual errors, instead of giving you the GC error.

**Comment by aj-r** [comment]
@thorn0 I have updated based on your comments. Please review again.

**Comment by thorn0** [comment]
Great. Let's introduce an alias for `boolean | null | undefined`. How about `Boolish`?

**Comment by aj-r** [comment]
@thorn0 done

**Review by thorn0 — APPROVED** [review]

**Comment by andy-ms** [comment]
:-1: for `boolean | null | undefined` , since the actual function does not look for `null | undefined` values -- it just treats the input as a boolean. Users who don't want to deal with nullness errors yet could just disable `--strictNullChecks`.

File: types/lodash/index.d.ts
Comment (rbuckton) [inline types/lodash/index.d.ts]: This seems like an odd choice if `T1` and `T2` are not comparable (e.g. when `comparator` is `undefined`). I would recommend you leave the original overload and add this one with both `values` and `comparator` as required, otherwise the entire method is much less type safe.

**Review by rbuckton — CHANGES_REQUESTED** [review]
I'm concerned about the loss of type safety introduced with `Comparator2` and the changes to the type parameters in many of the methods.

**Comment by RyanCavanaugh** [comment]
@aj-r Please address comments from the code reviewers.

**Comment by RyanCavanaugh** [comment]
Approved by a listed owner. PR ready to merge pending express review by a maintainer.

**Comment by RyanCavanaugh** [comment]
@rbuckton - Thanks for your review of this PR! Can you please look at the new code and update your review status if appropriate?

**Comment by aj-r** [comment]
@rbuckton Can you review this again now that I have made your requested changes?

**Comment by DanielRosenwasser** [comment]
Thanks for being patient!

**Pull request merged**
